src/Http/Http.Extensions/src/HttpContextJsonExtensions.Request.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http.Json
+{
+    public static partial class HttpContextJsonExtensions
+    {
+        public static ValueTask<TValue> ReadFromJsonAsync<TValue>(
+            this HttpRequest request,
+            CancellationToken cancellationToken = default)
+        {
+            return request.ReadFromJsonAsync<TValue>(options: null, cancellationToken);
+        }
+
+        public static ValueTask<TValue> ReadFromJsonAsync<TValue>(
+            this HttpRequest request,
+            JsonSerializerOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            if (!request.HasJsonContentType())
+            {
+                return new ValueTask<TValue>(Task.FromException<TValue>(CreateContentTypeError(request)));
+            }
+
+            return JsonSerializer.DeserializeAsync<TValue>(request.Body, options, cancellationToken);
+        }
+
+        public static ValueTask<object> ReadFromJsonAsync(
+            this HttpRequest request,
+            Type type,
+            CancellationToken cancellationToken = default)
+        {
+            return request.ReadFromJsonAsync(type, options: null, cancellationToken);
+        }
+
+        public static ValueTask<object> ReadFromJsonAsync(
+            this HttpRequest request,
+            Type type,
+            JsonSerializerOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            if (!request.HasJsonContentType())
+            {
+                return new ValueTask<object>(Task.FromException<object>(CreateContentTypeError(request)));
+            }
+
+            return JsonSerializer.DeserializeAsync(request.Body, type, options, cancellationToken);
```

Comment:
This needs to use a transcoding stream if the content encoding is not utf8. https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonInputFormatter.cs#L121-L130